horovod/torch/__init__.py

```diff
@@ -75,21 +77,28 @@ def hook(*ignore):
             assert p not in self._handles
             assert not p.grad.requires_grad
             name = self._parameter_names.get(p)
-            handle = allreduce_async_(p.grad.data, average=True, name=name)
-            self._handles[p] = handle
+
+            tensor = p.grad.data
+            compressor = self._compression.get_compressor(tensor)
+            tensor_compressed = compressor.compress(tensor)
+
+            handle = allreduce_async_(tensor_compressed, average=True, name=name)
+            self._handles[p] = (handle, compressor)
         return hook
 
     def synchronize(self):
-        for handle in self._handles.values():
-            synchronize(handle)
+        for p, value in self._handles.items():
+            handle, compressor = value
+            output = synchronize(handle)
+            p.grad.data = compressor.decompress(output)
```

Comment:
I wonder if we should do `.set_()` instead of assignment.  Should look into pros/cons if any.

Reply:
I was wondering about that myself.  For the non-fp16 case, this will be a no-op, as it'll just assign the tensor to itself.  But in the case of fp16, we always need to create a new tensor, because there is no way to do an in-place type cast.  

It may be possible to call `set_()` and pass in a tensor of a different type (fp16).  According to the documentation on [`set_()`](https://pytorch.org/docs/stable/tensors.html), the two tensors will share the same underlying storage, which seems problematic if the current type is fp32, but the parameter tensor type is fp16.

Reply:
Nah, we'd still need to convert, but I'm curious about tradeoff between `.set_()` vs `=`.  I'm wondering if any other thing may have a pointer to `p.grad.data`.

Reply:
I managed to track the implementation down to [here](https://github.com/torch/torch7/blob/master/lib/TH/generic/THTensor.c#L765) in Torch7 (I believe ATen is similar).

The difference, as I can tell, is that `set_` will delete the underlying storage of the previous tensor data, and set it to the provided tensor's storage (setting it for retention).  Whereas `=` will just replace the pointers, so any properties of the old tensor will be lost.

Overall, I think using `set_` is the safest option here.  I don't think it will be any faster, necessarily (though cleanup will happen faster, avoiding garbage collection).